apis/duck/v1alpha1/scalable_source_types.go

```diff
@@ -0,0 +1,235 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+	"knative.dev/pkg/apis/duck/v1"
+	"math"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"knative.dev/pkg/ptr"
+
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/apis/duck"
+)
+
+// KedaSource is an Implementable "duck type".
+var _ duck.Implementable = (*KedaSource)(nil)
+
+// SourceScaler specifies the minimum set of options that any scalable Source should support.
+type SourceScaler struct {
+	// MinScale defines the minimum scale for the source.
+	// If not specified, defaults to zero.
+	// +optional
+	MinScale *int32 `json:"minScale,omitempty"`
+
+	// MaxScale defines the maximum scale for the source.
+	// If not specified, defaults to one.
+	// +optional
+	MaxScale *int32 `json:"maxScale,omitempty"`
+}
+
+const (
+	// SourceScalerAnnotationKey is the annotation for the explicit class of
+	// source scaler that a particular resource has opted into. For example,
+	// sources.knative.dev/scaler: foo
+	SourceScalerAnnotationKey = "sources.knative.dev/scaler"
+)
+
+// +genduck
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// KedaSource is the minimum resource shape to adhere to in order to construct
+// a scalable Source using Keda (https://github.com/kedacore/keda).
+// This duck type is intended to allow implementors of KedaSources to verify
+// their own resources meet the expectations.
+// This is not a real resource.
+type KedaSource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   KedaSourceSpec   `json:"spec"`
+	Status KedaSourceStatus `json:"status"`
+}
+
+type KedaSourceSpec struct {
```

Comment:
why is the `Keda`?  --> Because it has keda-specific configuration

`why Source` --> because it implements the Source duck type.

why not just `Scalable` --> because is hard to do a generic `Scalable` one, different implementations have different configs.